MigrationsSqlGenerator: DefaultValue formats default value properly

mirol-h opened the pull request:
Fix for issue #14457. Contains breaking change in MigrationsSqlGenerator (changed signature of DefaultValue method).

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/EntityFrameworkCore?pullRequest=16909) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/EntityFrameworkCore?pullRequest=16909) before we can accept your contribution.<br/><br/>:x: mirol-h [sign now](https://cla.dotnetfoundation.org/aspnet/EntityFrameworkCore?pullRequest=16909)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/EntityFrameworkCore?pullRequest=16909) it.</sub>

ajcvickers commented on src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs:
```diff
@@ -1240,7 +1243,9 @@ protected virtual IUpdateSqlGenerator SqlGenerator
             else if (defaultValue != null)
             {
-                var typeMapping = Dependencies.TypeMappingSource.GetMappingForValue(defaultValue);
+                var typeMapping = type != null ?
```
Here we should try using the store type name first, but then fall back to the CLR type anyway if that fails, since in some cases on some providers we may not be able to interpret the store type name.

mirol-h commented:
Thanks for the feedback. I've incorporated your suggestion and also changed the call to `FindMapping` like this:

    Dependencies.TypeMappingSource.FindMapping(defaultValue.GetType(), type)

This solves the issue for SQLite provider, where this test failed: 

https://github.com/aspnet/EntityFrameworkCore/blob/c2a39d700963f71c0035fa297aae6ec430578951/test/EFCore.Sqlite.FunctionalTests/SqliteMigrationSqlGeneratorTest.cs#L50

    [ConditionalFact]
    public virtual void DefaultValue_formats_literal_correctly()
    {
        Generate(
            new CreateTableOperation
            {
                Name = "History",
                Columns =
                {
                    new AddColumnOperation
                    {
                        Name = "Event",
                        ClrType = typeof(string),
                        ColumnType = "TEXT",
                        DefaultValue = new DateTime(2015, 4, 12, 17, 5, 0)
                    }
                }
            });

        AssertSql(
            @"CREATE TABLE ""History"" (
	""Event"" TEXT NOT NULL DEFAULT '2015-04-12 17:05:00'
	);
	");
    }

It tried to generate SQL literal for TEXT from DateTime and failed because of cast from DateTime to string (I guess this is necessary (for SQLite), because SQLite lacks proper date time type, right?).

mirol-h commented:
@bricelam I've rebased branch onto release/3.0 and now lots of tests in SqlServer.FunctionalTests fail (e.g. `Query.AsTrackingSqlServerTest.Applied_to_projection` or `Query.QueryNavigationsSqlServerTest.Project_single_scalar_value_subquery_in_query_with_optional_navigation_works`).

Did I do something wrong, or is it something else?

bricelam commented:
Almost certainly unrelated. (Re-running the CI)

mirol-h commented:
@bricelam It's been some time, since I've rebased the branch on top of release/3.0. What's the current state of this PR (and #16960)? Is anything else required from me? In terms of 3.0 release, what are the odds at the moment that these PRs get merged?

bricelam commented:
You’re good. I’m just slow. I’ll get it into the 3.0.0 release.
